cgi-bin/DW/Controller/Search/Journal.pm

```diff
@@ -0,0 +1,185 @@
+    return error_ml( '/search.tt.error.wrongoffset' ) if $offset < 0 || $offset > 1000;
+
+    # we can GET a user name, but we can't use the controller's specify_user
+    # convenience method for this, because it will always default to remote,
+    # and that will override selection of site search
+    my $su = LJ::load_user( $get_args->{user} );  # may be overridden by POST
+    return error_ml( 'error.invaliduser' ) if $get_args->{user} && ! $su;
+
+    # form processing
+    if ( $r->did_post ) {
+        my $post_args = $r->post_args;
+
+        $su = LJ::load_user( $post_args->{mode} ) if $post_args->{mode};
+
+        # if no $su, then this is a public search, that's allowed.  but if it's a user,
+        # ensure that it's an account that we CAN search
+        $errors->add( "mode", ".error.forbidden" )
+            if $su && ! $su->allow_search_by( $remote );
+
+        # make sure we got a query
+        $q = LJ::strip_html( LJ::trim( $post_args->{query} ) );
+        $errors->add( "query", ".error.noquery" ) unless $q;
+        $errors->add( "query", ".error.longquery" ) if $q && length $q > 255;
+
+        my $sby = $post_args->{sort_by} || 'new';
+        $sby = 'new' unless $sby =~ /^(?:new|old|rel)$/;
+
+        my $wc = $post_args->{with_comments} ? 1 : 0;
+
+        $rv->{sort_by} = $sby;
+        $rv->{wc} = $wc;
+
+        # at this point, we should have good form data; the form can still
+        # have errors below, but they aren't the fault of the user input
+
+        unless ( $errors->exist ) {
+            # we have to set a few flags on what to search.  default to public and no bits.
+            my ( $ignore_security, $allowmask ) = ( 0, 0 );
+            if ( $su ) {
+                # if it's you, all posts, all bits
+                if ( $remote->equals( $su ) ) {
+                    $ignore_security = 1;
+
+                } elsif ( $su->is_community ) {
+                    # if it's a community you administer, also all bits
+                    if ( $remote->can_manage( $su ) ) {
+                        $ignore_security = 1;
+
+                    # for communities, member_of is the same as allow mask (no custom groups)
+                    } else {
+                        $allowmask = $remote->member_of( $su );
+                    }
+
+                # otherwise, if they trust you, get the mask ...
+                } elsif ( $su->trusts( $remote ) ) {
+                    $allowmask = $su->trustmask( $remote );
+                }
+            }
+
+            # the arguments to the search (userid=0 implies global search)
+            my $args = { userid => $su ? $su->id : 0, remoteid => $remote->id,
+                         query => $q, offset => $offset, sort_by => $sby,
+                         ignore_security => $ignore_security, allowmask => $allowmask,
+                         include_comments => $wc };
+
+            my $result = _do_search( Storable::nfreeze( $args ) );
```

Comment:
To clarify, you want me to change this to `_do_search( $args )` and have the `nfreeze` happen inside the function?